reconcile/utils/runtime/sharding.py

```diff
@@ -0,0 +1,509 @@
+                ):
+                    new_shard.shard_spec_overrides.sub_sharding = None
+                new_shard.shard_id = str(s)
+                new_shard.shards = str(num_shards)
+                new_shard.shard_name_suffix += f"-{s}"
+                shards.append(new_shard)
+            return shards
+
+
+class AWSAccountShardingStrategy:
+    IDENTIFIER = "per-aws-account"
+
+    def __init__(
+        self,
+        aws_accounts: Optional[Iterable[sharding_aws_accounts.AWSAccountV1]] = None,
+    ):
+
+        if not aws_accounts:
+            self.aws_accounts = (
+                sharding_aws_accounts.query(query_func=gql.get_api().query).aws_accounts
+                or []
+            )
+        else:
+            self.aws_accounts = list(aws_accounts)
+
+    def filter_accounts(
+        self, integration: str
+    ) -> list[sharding_aws_accounts.AWSAccountV1]:
+        return [
+            a
+            for a in self.aws_accounts
+            if (
+                not a.disable
+                or not a.disable.integrations
+                or (
+                    a.disable.integrations and integration not in a.disable.integrations
+                )
+            )
+        ]
+
+    def get_shard_spec_overrides(
+        self, sharding: Optional[IntegrationShardingV1]
+    ) -> dict[str, AWSAccountShardSpecOverrideV1]:
+
+        spos: dict[str, AWSAccountShardSpecOverrideV1] = {}
+
+        if isinstance(sharding, AWSAccountShardingV1) and sharding.shard_spec_overrides:
+            for sp in sharding.shard_spec_overrides:
+                spos[sp.shard.name] = sp
+        return spos
+
+    def check_integration_sharding_params(self, meta: IntegrationMeta) -> None:
+        if "--account-name" not in meta.args:
+            raise ValueError(
+                f"integration {meta.name} does not support the provided argument. "
+                " --account-name is required by the 'per-aws-account' sharding "
+                "strategy."
+            )
+
+    def build_shard_spec(
+        self,
+        aws_account: sharding_aws_accounts.AWSAccountV1,
+        integration_spec: IntegrationSpecV1,
+        spo: Optional[AWSAccountShardSpecOverrideV1],
+    ) -> ShardSpec:
+
+        return ShardSpec(
+            shard_key=aws_account.name,
+            shard_name_suffix=f"-{aws_account.name}",
+            extra_args=(integration_spec.extra_args or "")
+            + f" --account-name {aws_account.name}",
+            shard_spec_overrides=spo,
+        )
+
+    def build_integration_shards(
+        self,
+        integration_meta: IntegrationMeta,
+        integration_managed: IntegrationManagedV1,
+    ) -> list[ShardSpec]:
+
+        self.check_integration_sharding_params(integration_meta)
+        spos = self.get_shard_spec_overrides(integration_managed.sharding)
+        shards = []
+        for c in self.filter_accounts(integration_meta.name):
```

Comment:
I think, if you'd pass in this iterable into the `build_integration_shards`, this method could be reused for all sharding strategies. Could also make it easier to add sub sharding. 
This would require some base class for the ShardingStrategies.

Reply:
My first intention was to have a base class for all the key-based sharding strategies (it's commented at the end for the sharding file), but I discarded it because the `ShardOverrideClasses` can not be generated with an interface. At that time, I wanted to avoid having to type every new Override in the base class method...

But it makes sense, so I'll try to have a base class with less logic.

Reply:
I discarded the idea because build_integration_shards interacts with the overrides classes, and I don't fully like to type all of them in the base class logic. The override classes are "generated", and as they are already part of an interface (integrationSharding), they can not be grouped in another interface.

I know this is complex and is not clear enough, I can elaborate more if you want outside of the PR.